File: src/vehicle_sim/src/include/TireConfig.cpp
#include "TireConfig.hpp"
#include <cmath> // for std::abs, std::min, std::sqrt

TireConfig::TireConfig()
    : B(0.7), C(1.5), D(1.2), E(0.8), f(100.0), coeff_rr(0.008) {} //
TireConfig::TireConfig(double B_, double C_, double D_, double E_, double f_, double coeff_rr_)
    : B(B_), C(C_), D(D_), E(E_), f(f_), coeff_rr(coeff_rr_) {}

// tire forces
void TireConfig::setTireParams(const double B_, const double C_, const double D_, const double E_, const double f_)
{
    B = B_;
    C = C_;
    D = D_;
    E = E_;
    f = f_;
}
void TireConfig::getTireParams(double &B_, double &C_, double &D_, double &E_, double &f_) const
{
    B_ = B;
    C_ = C;
    D_ = D;
    E_ = E;
    f_ = f;
}

void TireConfig::setCoeffRR(const double coeff_rr_)
{
    coeff_rr = coeff_rr_;
}
double TireConfig::getCoeffRR() const
{
    return coeff_rr;
}

void TireConfig::calcTireForces(const double slip_angle, const double slip_ratio, const double w_wheel, const double Fz_wheel, double &Fx_wheel, double &Fy_wheel)
{
    // slip_angle [deg]
    // slip_ratio [ratio]
    double slip_angle_corr = slip_angle * std::min(1.0, std::abs(w_wheel) * 0.01);
    double slipNet = std::sqrt(slip_angle_corr * slip_angle_corr + slip_ratio * slip_ratio * f * f);
    if (slipNet < 1e-9)
    {
        Fx_wheel = 0.0;
        Fy_wheel = 0.0;
        return;
    }

    // pacejka tire formula
    double Fnet = D * std::sin(C * std::atan(B * slipNet - E * (B * slipNet - std::atan(B * slipNet))));

    // calculate forces in each direction
    double F_rolling_resistance = coeff_rr * Fz_wheel;
    double F_partial = Fz_wheel * Fnet / slipNet;
    Fx_wheel = F_partial * (slip_ratio * f) - F_rolling_resistance;
    Fy_wheel = -F_partial * slip_angle_corr; // * std::min(1.0, std::abs(w_wheel)*10+0.01);
}

File: matlab/fcn/TireConfig.m
classdef TireConfig
    properties
        B
        C
        D
        E
        f
        coeff_rr
        coeff_stiction
    end

    methods
        function obj = TireConfig(s)
            if strcmp(s, 'sym')
                for fn = string(fieldnames(obj))'
                    obj.(fn) = sym(fn);
                end
            else
                obj.B = s.B;
                obj.C = s.C;
                obj.D = s.D;
                obj.E = s.E;
                obj.f = s.f;
                obj.coeff_rr = s.coeff_rr;
                obj.coeff_stiction = s.coeff_stiction;
            end
        end

        function [Fx_wheel, Fy_wheel] = calcTireForces(obj, slip_angle, slip_ratio, w_wheel, Fz_wheel)
            if isa(slip_angle, 'sym')
                slip_angle_corr = slip_angle;
            else
                slip_angle_corr = slip_angle * min(1, abs(w_wheel)*0.01);
            end
            slipNet = sqrt(slip_angle_corr^2 + slip_ratio^2 * obj.f^2);
            if ~isa(slipNet, 'sym') && slipNet < 1e-9
                Fx_wheel = 0;
                Fy_wheel = 0;
            else
                Fnet = obj.D * sin(obj.C * atan(obj.B*slipNet - obj.E * (obj.B * slipNet - atan(obj.B * slipNet))));
                F_rolling_resistance = obj.coeff_rr * Fz_wheel;
                if isa(slip_angle, 'sym')
                    F_rolling_stiction = obj.coeff_stiction * Fz_wheel * sign(w_wheel);
                else
                    F_rolling_stiction = obj.coeff_stiction * Fz_wheel * sign(w_wheel) * min(abs(w_wheel)*100, 1);
                end
                F_partial = Fz_wheel * Fnet / slipNet;
                Fx_wheel = F_partial * (slip_ratio * obj.f) - F_rolling_resistance - F_rolling_stiction;
                Fy_wheel = - F_partial * slip_angle_corr;
            end
        end
    end
end